File: src/y2024/Day19.kt
package y2024

import kotlin.time.measureTimedValue

fun main() {

    fun parseInput(input: List<String>) =
        input.first().split(", ").toSet().let { it to it.maxOf { it.length } }

    fun part1(input: List<String>) = run {
        val (towels, maxLength) = parseInput(input)
        val map = mutableMapOf<String, Boolean>()

        val rec = DeepRecursiveFunction { s: String ->
            s.isEmpty() || map.getOrPut(s) {
                (1 .. s.length.coerceAtMost(maxLength))
                    .filter { s.take(it) in towels }
                    .any { callRecursive(s.drop(it)) }
            }
        }

        input.subList(2, input.size).count { rec(it) }
    }

    fun part2(input: List<String>) = run {
        val (towels, maxLength) = parseInput(input)
        val map = mutableMapOf<String, Long>()

        val rec = DeepRecursiveFunction { s: String ->
            if(s.isEmpty()) 1L else map.getOrPut(s) {
                (1 .. s.length.coerceAtMost(maxLength))
                    .filter { s.take(it) in towels }
                    .sumOf { callRecursive(s.drop(it)) }
            }
        }

        input.subList(2, input.size).sumOf { rec(it) }
    }

    val testInput = readInput("Day19_test")
    check(part1(testInput).apply { println() } == 6)
    check(part2(testInput).apply { println() } == 16L)

    val input = readInput("Day19")
    measureTimedValue { part1(input) }.println()
    measureTimedValue { part2(input) }.println()
}

File: src/y2024/Day19.py
from functools import cache

def dfs(op, towels, max_length):
    fun = cache(lambda s: 1 if s == "" else op(fun(s[i:]) for i in range(1, min(max_length, len(s)) + 1) if s[:i] in towels))
    return fun

def parse_input(input):
    input = list(map(str.strip, input))
    towels = set(input[0].split(", "))
    return input, towels, max(map(len, towels))

def solve(input, op):
    input, towels, max_length = parse_input(input)
    return sum(map(dfs(op, towels, max_length), input[2:]))

def part1(input):
    return solve(input, any)

def part2(input):
    return solve(input, sum)

test_input = open("Day19_test.txt", "r").readlines()
input = open("Day19.txt", "r").readlines()

p1 = part1(test_input)
p2 = part2(test_input)
assert p1 == 6
print(p1)
assert p2 == 16
print(p2)

print(part1(input))
print(part2(input))